clippy_utils/src/ty/mod.rs

```diff
@@ -1376,3 +1376,62 @@ pub fn option_arg_ty<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> Option<Ty<'t
         _ => None,
     }
 }
+
+/// Check if a Ty<'_> of `Iterator` has side effects when iterated over by checking if it
+/// captures any mutable references or equivalents.
+pub fn is_iter_with_side_effects<'tcx>(cx: &LateContext<'tcx>, iter_ty: Ty<'tcx>) -> bool {
+    let Some(iter_trait) = cx.tcx.lang_items().iterator_trait() else {
+        return false;
+    };
+
+    is_iter_with_side_effects_impl(cx, iter_ty, iter_trait)
+}
+
+fn is_iter_with_side_effects_impl<'tcx>(cx: &LateContext<'tcx>, iter_ty: Ty<'tcx>, iter_trait: DefId) -> bool {
+    if let ty::Adt(adt_def, args) = iter_ty.kind() {
+        return adt_def
+            .all_fields()
+            .any(|field| is_ty_with_side_effects(cx, field.ty(cx.tcx, args), iter_trait));
+    }
+
+    false
+}
+
+fn is_ty_with_side_effects<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>, iter_trait: DefId) -> bool {
+    match ty.kind() {
+        ty::RawPtr(..) | ty::Ref(..) | ty::Adt(..) => is_mutable_reference_or_equivalent_and(cx, ty, |inner_ty| {
+            !implements_trait(cx, inner_ty, iter_trait, &[]) || is_iter_with_side_effects_impl(cx, inner_ty, iter_trait)
+        }),
+        ty::Closure(_, closure_args) => {
+            matches!(closure_args.types().next_back(), Some(captures) if captures.tuple_fields().iter().any(|capture_ty| is_ty_with_side_effects(cx, capture_ty, iter_trait)))
+        },
+        ty::Array(elem_ty, _) | ty::Slice(elem_ty) => is_ty_with_side_effects(cx, *elem_ty, iter_trait),
+        ty::Tuple(field_tys) => field_tys
+            .iter()
+            .any(|field_ty| is_ty_with_side_effects(cx, field_ty, iter_trait)),
+        _ => false,
+    }
+}
+
+/// Check if `ty` is a mutable reference or equivalent. This includes:
+/// - A mutable reference/pointer.
+/// - A reference/pointer to a non-`Freeze` type.
+/// - A `PhantomData` type containing any of the previous.
+pub fn is_mutable_reference_or_equivalent<'tcx>(cx: &LateContext<'tcx>, ty: Ty<'tcx>) -> bool {
+    is_mutable_reference_or_equivalent_and(cx, ty, |_| true)
+}
+
+fn is_mutable_reference_or_equivalent_and<'tcx, F>(cx: &LateContext<'tcx>, ty: Ty<'tcx>, pred: F) -> bool
+where
+    F: Fn(Ty<'tcx>) -> bool,
+{
+    match ty.kind() {
+        ty::RawPtr(ty, mutability) | ty::Ref(_, ty, mutability) => {
+            (mutability.is_mut() || !ty.is_freeze(cx.tcx, cx.typing_env())) && pred(*ty)
+        },
+        ty::Adt(adt_def, args) => adt_def.all_fields().any(|field| {
+            matches!(field.ty(cx.tcx, args).kind(), ty::Adt(adt_def, args) if adt_def.is_phantom_data() && args.types().any(|arg_ty| is_mutable_reference_or_equivalent(cx, arg_ty)))
+        }),
+        _ => false,
+    }
+}
```

Comment:
These need to be merged. ADTs, closures, tuples etc. need to be treated the same at every level of nesting.